supervisor feature switch states replications on workload namespaces

divyenpatel opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
Following changes are done in this PR

- Added a service to replicate supervisor feature switch states to workload namespaces.
- when service is started, it is creating or updating `cnscsisvfeaturestate` CR in all workload namespaces.
- any failure to create or update CR will be re-tried at regular intervals.
- service is watching on FSS config-map changes on the supervisor and updates all workload namespaces with the latest feature switch states. 
- service is watching on new workload namespaces added on the cluster and helps create `cnscsisvfeaturestate` CR with the latest feature switch states.
- service will crash when we observe that the fss config-map is deleted or not found or we fail to parse the config-map value.


**Special notes for your reviewer**:
Testing done

Verified CR is registered and created by `SvFSSReplicationService`.

```
# kubectl api-resources -o wide | grep cnscsisvfeaturestates
cnscsisvfeaturestates                                cns.vmware.com/v1alpha1                      true         CnsCsiSvFeatureStates             [delete deletecollection get list patch create update watch]
```

Verified CR is created in the workload namespace

```
# kubectl get cnscsisvfeaturestates --all-namespaces
NAMESPACE             NAME              AGE
test-gc-e2e-demo-ns   svfeaturestates   27s

# kubectl describe cnscsisvfeaturestates/svfeaturestates --namespace=test-gc-e2e-demo-ns
Name:         svfeaturestates
Namespace:    test-gc-e2e-demo-ns
Labels:       <none>
Annotations:  <none>
API Version:  cns.vmware.com/v1alpha1
Kind:         CnsCsiSvFeatureStates
Metadata:
  Creation Timestamp:  2021-03-26T07:38:59Z
  Generation:          1
  Managed Fields:
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:featureStates:
    Manager:         vsphere-syncer
    Operation:       Update
    Time:            2021-03-26T07:38:59Z
  Resource Version:  326666
  Self Link:         /apis/cns.vmware.com/v1alpha1/namespaces/test-gc-e2e-demo-ns/cnscsisvfeaturestates/svfeaturestates
  UID:               36772301-12fc-44d8-8e9c-3606cb36bf73
Spec:
  Feature States:
    Enabled:  true
    Name:     volume-extend
    Enabled:  true
    Name:     volume-health
    Enabled:  false
    Name:     vsan-direct-disk-decommission
    Enabled:  false
    Name:     csi-auth-check
    Enabled:  true
    Name:     csi-sv-feature-states-replication
    Enabled:  false
    Name:     file-volume
    Enabled:  true
    Name:     online-volume-extend
Events:       <none>
```

Updated `csi-feature-states` config-map in the `vmware-system-csi` namespace and added new-feature: 'true' and verified new state is updated in CRs.

```
# kubectl describe cnscsisvfeaturestates/svfeaturestates --namespace=test-gc-e2e-demo-ns
Name:         svfeaturestates
Namespace:    test-gc-e2e-demo-ns
Labels:       <none>
Annotations:  <none>
API Version:  cns.vmware.com/v1alpha1
Kind:         CnsCsiSvFeatureStates
Metadata:
  Creation Timestamp:  2021-03-26T07:38:59Z
  Generation:          26
  Managed Fields:
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:featureStates:
    Manager:         vsphere-syncer
    Operation:       Update
    Time:            2021-03-26T07:38:59Z
  Resource Version:  630156
  Self Link:         /apis/cns.vmware.com/v1alpha1/namespaces/test-gc-e2e-demo-ns/cnscsisvfeaturestates/svfeaturestates
  UID:               36772301-12fc-44d8-8e9c-3606cb36bf73
Spec:
  Feature States:
    Enabled:  true
    Name:     csi-sv-feature-states-replication
    Enabled:  false
    Name:     file-volume
    Enabled:  true
    Name:     new-feature
    Enabled:  true
    Name:     online-volume-extend
    Enabled:  true
    Name:     volume-extend
    Enabled:  true
    Name:     volume-health
    Enabled:  false
    Name:     vsan-direct-disk-decommission
    Enabled:  false
    Name:     csi-auth-check
Events:       <none>
```

Test logs with latest commit on this PR.

```
# kubectl logs vsphere-csi-controller-778f8c6469-829ft --namespace=vmware-system-csi -c vsphere-syncer | grep "featurestates/featurestates"
{"level":"info","time":"2021-04-01T17:23:02.602032018Z","caller":"featurestates/featurestates.go:79","msg":"Starting SvFSSReplicationService"}
{"level":"info","time":"2021-04-01T17:23:03.247089098Z","caller":"featurestates/featurestates.go:407","msg":"successfully retrieved SV feature switch state from namespace:\"vmware-system-csi\" with name: \"csi-feature-states\""}
{"level":"info","time":"2021-04-01T17:23:03.257189554Z","caller":"featurestates/featurestates.go:259","msg":"observed fss add: fss: map[csi-auth-check:false csi-sv-feature-states-replication:true file-volume:false online-volume-extend:true volume-extend:true volume-health:true vsan-direct-disk-decommission:false]","TraceId":"6c2b76d1-dcf6-414c-b9ba-85fa495a1518"}
{"level":"info","time":"2021-04-01T17:23:03.28052837Z","caller":"featurestates/featurestates.go:345","msg":"observed new workload namespace: test-gc-e2e-demo-ns","TraceId":"d8c99350-f29d-48fd-9957-9e55f90937ca"}
{"level":"info","time":"2021-04-01T17:23:03.280665151Z","caller":"featurestates/featurestates.go:242","msg":"enqueued CR updates for workload namespace: \"test-gc-e2e-demo-ns\"","TraceId":"d8c99350-f29d-48fd-9957-9e55f90937ca"}
{"level":"info","time":"2021-04-01T17:23:03.362006548Z","caller":"featurestates/featurestates.go:154","msg":"informer on config-map and namespaces started"}
{"level":"info","time":"2021-04-01T17:23:03.362278406Z","caller":"featurestates/featurestates.go:158","msg":"started background routine to process pending feature state updates at regular interval"}
{"level":"info","time":"2021-04-01T17:23:03.362380642Z","caller":"featurestates/featurestates.go:159","msg":"SvFSSReplicationService is running"}
{"level":"info","time":"2021-04-01T17:23:33.363806664Z","caller":"featurestates/featurestates.go:179","msg":"feature state update is required for namespace: \"test-gc-e2e-demo-ns\"","TraceId":"27d590a9-f5ad-4e46-8478-cf3df3d5e9cf"}
{"level":"info","time":"2021-04-01T17:23:33.370967476Z","caller":"featurestates/featurestates.go:196","msg":"cnsCsiSvFeatureStates CR instance is not present in the namespace: \"test-gc-e2e-demo-ns\". Creating CR with latest feature switch state, Err: cnscsisvfeaturestates.cns.vmware.com \"svfeaturestates\" not found","TraceId":"27d590a9-f5ad-4e46-8478-cf3df3d5e9cf"}
{"level":"info","time":"2021-04-01T17:23:33.399092038Z","caller":"featurestates/featurestates.go:212","msg":"created cnsCsiSvFeatureStates CR instance in the namespace: \"test-gc-e2e-demo-ns\"","TraceId":"27d590a9-f5ad-4e46-8478-cf3df3d5e9cf"}
{"level":"info","time":"2021-04-01T17:24:45.684198506Z","caller":"featurestates/featurestates.go:298","msg":"observed fss update: oldfss: map[csi-auth-check:false csi-sv-feature-states-replication:true file-volume:false online-volume-extend:true volume-extend:true volume-health:true vsan-direct-disk-decommission:false] newfss: map[csi-auth-check:false csi-sv-feature-states-replication:true file-volume:false online-volume-extend:true test1:true test2:false volume-extend:true volume-health:true vsan-direct-disk-decommission:false]","TraceId":"77f98051-f49f-4a37-a404-472a83a2aed1"}
{"level":"info","time":"2021-04-01T17:24:45.744059853Z","caller":"featurestates/featurestates.go:231","msg":"enqueued CR updates for workload namespace: \"test-gc-e2e-demo-ns\"","TraceId":"77f98051-f49f-4a37-a404-472a83a2aed1"}
{"level":"info","time":"2021-04-01T17:25:03.367747416Z","caller":"featurestates/featurestates.go:174","msg":"starting to process pending feature state updates","TraceId":"3c22daf6-1ac4-44d6-bc75-1a81e99d954e"}
{"level":"info","time":"2021-04-01T17:25:03.367821945Z","caller":"featurestates/featurestates.go:179","msg":"feature state update is required for namespace: \"test-gc-e2e-demo-ns\"","TraceId":"3c22daf6-1ac4-44d6-bc75-1a81e99d954e"}
{"level":"info","time":"2021-04-01T17:25:03.408830725Z","caller":"featurestates/featurestates.go:193","msg":"updated cnsCsiSvFeatureStates CR instance in the namespace: \"test-gc-e2e-demo-ns\"","TraceId":"3c22daf6-1ac4-44d6-bc75-1a81e99d954e"}
```


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
supervisor feature switch states replications on workload namespaces
```

shalini-b commented on manifests/dev/vsphere-7.0u2/supervisorcluster/1.19/vsphere-csi-controller.yaml:
```diff
@@ -47,6 +47,12 @@ rules:
   resources: ["triggercsifullsyncs"]
   verbs: ["create", "get", "update", "watch", "list"]
+- apiGroups: [""]
+  resources: ["namespace"]
```
Can we `namespace` on line 18 instead?

divyenpatel replied:
fixed.

shalini-b commented on pkg/kubernetes/informers.go:
```diff
@@ -99,6 +99,20 @@ func (im *InformerManager) AddPVListener(add func(obj interface{}), update func(
+	if im.namespaceInformer == nil {
+		im.namespaceInformer = im.informerFactory.Core().V1().Namespaces().Informer()
+	}
+	im.namespaceSynced = im.pvInformer.HasSynced
```
`pvInformer` -> `namespaceInformer`

divyenpatel replied:
fixed.

shalini-b commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,368 @@
+
+	// Start Retry routine to process failed create or update CRs
+	go pendingCRUpdatesObj.startRetryAttempts()
+	log.Infof("started background routine to process failed CR updates at regulat interval")
```
`regulat` -> `regular`

shalini-b commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,368 @@
+						if !apierrors.IsNotFound(err) {
+							// attempt to Create the CR
+							err = controllerRuntimeClient.Create(ctx, featurestatesSpec)
+						}
```
Can we have a log in `else` in case `err` is something other than NotFound?

divyenpatel replied:
logged error message

shalini-b commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,368 @@
+						}
+					} else {
+						// Attempt to Update CR
+						err = controllerRuntimeClient.Update(ctx, featurestatesSpec)
```
Can we remove the `else` and keep the update outside?

divyenpatel replied:
we need to keep update inside, as, after create, we do not want to apply update.

shalini-b commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,368 @@
+	//  Retrieve SV FeatureStates configmap
+	fssConfigMap, err := k8sClient.CoreV1().ConfigMaps(supervisorFeatureStateConfigMapNamespace).Get(ctx, supervisorFeatureStatConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		log.Errorf("failed to retrive SV featureswitch state from namespace:%q with name: %q", supervisorFeatureStateConfigMapNamespace, supervisorFeatureStatConfigMapName)
```
`featureswitch` -> `feature switch`

divyenpatel replied:
done.

shalini-b commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,368 @@
+		featureState.Enabled, err = strconv.ParseBool(state)
+		if err != nil {
+			log.Errorf("failed to parse feature state value: %v for feature: %q", state, feature)
+			return nil, err
```
As this function indirectly gets called in `StartSvFSSReplicationService`, should we error out if the value is not correct or just mark the feature as false? I guess our current code marks a feature state as false if the value is not bool

divyenpatel replied:
done. instead of erroring out, when parsing fails, marking the feature as false.

shalini-b commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,368 @@
+			if err == nil {
+				break
+			}
+			log.Errorf("failed to create feature states CR in the workload namespace: %q, err: %v", namespace, err)
```
Should we add a sleep here as you are retrying infinitely?

divyenpatel replied:
done

shalini-b commented on pkg/syncer/metadatasyncer.go:
```diff
@@ -168,6 +170,22 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
 	}
 
+	if metadataSyncer.clusterFlavor == cnstypes.CnsClusterFlavorWorkload {
+		if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.CSISVFeatureStateReplication) {
```
Don't we want to check if it is a seed-shoot cluster here? Are we planning to use this technique for TKGS in default as well?

divyenpatel replied:
no need to check the seed or shoot cluster here. we need to start FSS replication service in the supervisor csi controller pod.

chethanv28 commented on manifests/dev/vsphere-7.0u2/supervisorcluster/1.18/vsphere-csi-controller.yaml:
```diff
@@ -47,6 +47,12 @@ rules:
   resources: ["triggercsifullsyncs"]
   verbs: ["create", "get", "update", "watch", "list"]
+- apiGroups: [""]
+  resources: ["namespace"]
```
`namespaces` ? for consistency

chethanv28 replied:
How about moving this to Line: 18

divyenpatel replied:
done

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+	}
+
+	// create/update feature state CRs in all workload namespaces
+	err = updateFeatureStatesCRInAllWorkLoadNamespaces(ctx, []featurestatesv1alpha1.FeatureState{})
```
`updateFeatureStatesCRInAllWorkLoadNamespaces` ==> HOw about calling it `replicateFeatureStatesCRInAllWorkLoadNamespaces`

chethanv28 replied:
I think its good to pass featureStates pre-populated by adding below code before this line
```
featureStates, err := getFeatureStates(ctx, k8sClient)
	if err != nil {
		log.Errorf("failed to get latest featurestates. Err: %v", err)
		return err
	}
```

divyenpatel replied:
@chethanv28 I have made a major change in the code. please take a look at the latest diff.

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+	// Create k8s Informer and watch on configmaps and namespaces
+	informer := k8s.NewInformer(k8sClient)
+	informer.AddConfigMapListener(ctx, k8sClient, svFeatureStateConfigMapNamespace,
+		// Add
```
Can you confirm why we don't need a handler `Add` ? Wouldn't we need this at the time of configmap creation ?

SandeepPissay replied:
I have the same question.

divyenpatel replied:
I thought no one should be adding or removing config-map while SV controller is running. but if user chooses to delete and re-create config-map to add/update feature states, it is better to watch on add config-map.
I will address this.

divyenpatel replied:
refactored code watched on config-map add event

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+		LabelSelector: WorkLoadNamespaceLabelKey,
+	})
+	if err != nil {
+		log.Errorf("failed to get workload namespaces. Err: %v", err)
```
failed to get workload namespaces ==> `failed to list workload namespaces`

divyenpatel replied:
done.

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+	}
+}
+
+// configMapAdded adds feature state switch values from configmap that has been created on K8s cluster
```
update comment

divyenpatel replied:
done

```
// namespaceAdded adds is called when new namespace is added on the k8s cluster.
func namespaceAdded(obj interface{}) {
```

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+				log.Infof("created feature states CR in the namespaces: %q", namespace.Name)
+				break
+			}
+			log.Errorf("failed to create feature states CR in the workload namespace: %q, err: %v", namespace, err)
```
namespace.Name ?

divyenpatel replied:
done.

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+	//  Retrieve SV FeatureStates configmap
+	fssConfigMap, err := k8sClient.CoreV1().ConfigMaps(supervisorFeatureStateConfigMapNamespace).Get(ctx, supervisorFeatureStatConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		log.Errorf("failed to retrive SV feature switch state from namespace:%q with name: %q", supervisorFeatureStateConfigMapNamespace, supervisorFeatureStatConfigMapName)
```
retrive ==> retrieve

divyenpatel replied:
done

chethanv28 commented on pkg/kubernetes/informers.go:
```diff
@@ -99,6 +99,20 @@ func (im *InformerManager) AddPVListener(add func(obj interface{}), update func(
+// AddNamespaceListener hooks up add, update, delete callbacks
+func (im *InformerManager) AddNamespaceListener(add func(obj interface{}), update func(oldObj, newObj interface{}), remove func(obj interface{})) {
+	if im.namespaceInformer == nil {
+		im.namespaceInformer = im.informerFactory.Core().V1().Namespaces().Informer()
```
Have considered using `NewFilteredNamespaceInformer` with `TweakListOptionsFunc` to watch only on workload namespaces?

chethanv28 commented on pkg/syncer/metadatasyncer.go:
```diff
@@ -168,6 +170,22 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
 
+	if metadataSyncer.clusterFlavor == cnstypes.CnsClusterFlavorWorkload {
+		if metadataSyncer.coCommonInterface.IsFSSEnabled(ctx, common.CSISVFeatureStateReplication) {
+			log.Info("Starting SvFSSReplicationService")
```
You can move this to Line: 181

divyenpatel replied:
actually moved it inside the function StartSvFSSReplicationService

divyenpatel replied:
moved it inside - StartSvFSSReplicationService func

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+				log.Infof("updated feature states: %v in the all workload namespaces", featureStates)
+				break
+			} else {
+				log.Errorf("failed to update feature states CR in the all workload namespaces. err: %v", err)
```
this can be `failed to update the following workload namespaces: %v, pendingCRUpdatesObj.pendingCRUpdate` ?

divyenpatel replied:
updateFeatureStatesCRInAllWorkLoadNamespaces only returns error when it is not able to retrieve namespace or fail to parse featureswitch states from config-map. changed this to following message

```
log.Errorf("failed to update feature states CR in the workload namespaces. err: %v", err)
```

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+	}
+
+	// create/update feature state CRs in all workload namespaces
+	err = updateFeatureStatesCRInAllWorkLoadNamespaces(ctx, []featurestatesv1alpha1.FeatureState{})
```
Can we not handle Add events for the SV FSS config map listener  and move the initial creation of CRs in that method?

divyenpatel replied:
done listening on add config-map event.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+	log.Infof("started background routine to process failed CR updates at regular interval")
+	log.Infof("SvFSSReplicationService is running")
+	var stopCh = make(chan bool)
+	<-stopCh
```
What happens if we do not block?

divyenpatel replied:
if we do not block, then routine will end.

SandeepPissay replied:
I meant to ask, would that be a problem? And moreover we have a local channel here which means another go routine cannot stop this go routine if needed (ex: graceful shutdown of syncer).

SandeepPissay reviewed (commented):
Looks like the diff is updated and I cannot add any more comments. I will review the latest.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+			}
+			featureStates = append(featureStates, featureState)
+		}
+		for {
```
Why do we need this infinite loop? If `updateFeatureStatesCRInAllWorkLoadNamespaces` fails, we are adding it to the queue right?

divyenpatel replied:
because I do not want to miss the config-update. this retry here is for error other than create/update CR error.
failed to create or update CR is cached.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+	}
+	if _, ok = namespace.Labels[WorkLoadNamespaceLabelKey]; ok {
+		log.Infof("observed new workload namespace: %v", namespace.Name)
+		for {
```
Why do we have an infinite loop?

divyenpatel replied:
this is for handling errors other than create CR, when the latest feature switch state is not retrieved or we fail to parse feature switch state. added retry in the loop here, so I do not miss the namespace add event.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,399 @@
+					featurestateCR := &featurestatesv1alpha1.CnsCsiSvFeatureStates{}
+					err = controllerRuntimeClient.Get(ctx, client.ObjectKey{Name: SVFeatureStateCRName, Namespace: namespace}, featurestateCR)
+					if err != nil {
+						if !apierrors.IsNotFound(err) {
```
Shouldn't this be `if apierrors.IsNotFound(err)` so that you can create the CR is it is not found?

divyenpatel replied:
yes. corrected.

SandeepPissay reviewed (commented):
I see the problem as a producer-consumer problem. Producers(config map listener and namespace listener) add the SV namespace to the queue, and consumer creates the CR on all the namespaces in the queue. The only reason why the consumer may fail is when the API server is not responding. In this case, the consumer can just retry after few seconds. 

The current code queues not only the SV namespace, but also the FSS which could be stale, and we rely on confiMapUpdater method to update the FSS in the queue. We can avoid all this by having a simple algorithm where only the SV namespace is queued. Before the retries goes over the queue, it can fetch the latest SV FSS and use that for creating the CRs. I also think the current code is a bit complex.

divyenpatel commented:
> I see the problem as a producer-consumer problem. Producers(config map listener and namespace listener) add the SV namespace to the queue, and consumer creates the CR on all the namespaces in the queue. The only reason why the consumer may fail is when the API server is not responding. In this case, the consumer can just retry after few seconds.
> 
> The current code queues not only the SV namespace, but also the FSS which could be stale, and we rely on confiMapUpdater method to update the FSS in the queue. We can avoid all this by having a simple algorithm where only the SV namespace is queued. Before the retries goes over the queue, it can fetch the latest SV FSS and use that for creating the CRs. I also think the current code is a bit complex.

We will not hit into state FSS update with this approach, as any time when new update is applied, if we find pending update in the retry map, we mark that pending re-try task as processed, so it does not get picked up while processing retry for pending updates.

But yes as discussed offline this can be simplified so we do not need infinite loops in the callback in the config-map update/add and namespace add.  I am refactoring the code as per our discussion. I will ping when it is ready for review.

shalini-b commented on pkg/kubernetes/types.go:
```diff
@@ -56,6 +56,10 @@ type InformerManager struct {
 	configMapInformer cache.SharedInformer
+
+	// namespaceInformer informer
+	namespaceInformer cache.SharedInformer
```
Can we place `namespaceInformer` and `namespaceSynced` together? Just like every other resource in this struct

divyenpatel replied:
done

divyenpatel commented:
@SandeepPissay @chethanv28 @shalini-b I have refactored the code based on feedback I have received.
 In the previous commit, we were having blocking calls, multiple endless `for` loops and retry attempts in the config-map and namespace listener which is removed.

The major change here is, we are queuing all updates observed while watching on config-map add/change and namespace addition and processing them at regular intervals. I have addressed all your review comments and validate this change on the supervisor cluster. In the updated code, I have utilized config-map and namespace informers to keep up-to-date in-memory feature states and workload namespaces, so we do not need to query latest FSS and workload namespaces while creating/updating CR.

divyenpatel commented:
jtest all

svcbot-qecnsdp commented:
Started GC block pipeline...

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
GC build status: FAILURE 
Stage before exit: deploy-gc-testbed 

```

chethanv28 commented on manifests/dev/vsphere-7.0u2/supervisorcluster/1.17/vsphere-csi-controller.yaml:
```diff
@@ -302,6 +305,7 @@ data:
   "volume-extend": "true"
   "volume-health": "true"
   "trigger-csi-fullsync" : "false"
```
not related to this PR, remove space after `"trigger-csi-fullsync"`

chethanv28 commented on manifests/dev/vsphere-7.0u2/supervisorcluster/1.18/vsphere-csi-controller.yaml:
```diff
@@ -302,6 +305,7 @@ data:
   "volume-extend": "true"
   "volume-health": "true"
   "trigger-csi-fullsync" : "false"
```
Same as above

divyenpatel replied:
rebased PR to include yaml changes.

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,438 @@
+		return nil, err
+	}
+
+	log.Infof("sucessfully retrieved SV feature switch state from namespace:%q with name: %q", supervisorFeatureStateConfigMapNamespace, supervisorFeatureStatConfigMapName)
```
`sucessfully` ==> `successfully`

divyenpatel replied:
done

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,438 @@
+		featureState.Name = feature
+		featureState.Enabled, err = strconv.ParseBool(state)
+		if err != nil {
+			log.Errorf("failed to parse feature state value: %v for feature: %q. Defaulting to false", state, feature)
```
Looks like we ignore error here and move to next steps. Potentially will this cause any NPE ?
If ParseBool fails, don't we have to set `featureState.Enabled` = `false`.

divyenpatel replied:
If ParseBool fails it retuns false with error, so we are good.

```
// ParseBool returns the boolean value represented by the string.
// It accepts 1, t, T, TRUE, true, True, 0, f, F, FALSE, false, False.
// Any other value returns an error.
func ParseBool(str string) (bool, error) {
	switch str {
	case "1", "t", "T", "true", "TRUE", "True":
		return true, nil
	case "0", "f", "F", "false", "FALSE", "False":
		return false, nil
	}
	return false, syntaxError("ParseBool", str)
}
```

SandeepPissay replied:
I would suggest to crash CSI in such cases to surface such problems sooner. Ignoring errors is problematic as system continues to work with stale data and logs getting rolled over.

chethanv28 commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,438 @@
+			featureState.Name = feature
+			featureState.Enabled, err = strconv.ParseBool(state)
+			if err != nil {
+				log.Errorf("failed to parse feature state value: %v for feature: %q", state, feature)
```
Same comment as below, will this potentially cause any NPE if we ignore errors. Can you set `featureState.Enabled` = `false`

divyenpatel replied:
If ParseBool fails it returns false with error, so we are good.

chethanv28 reviewed (changes requested):
Can you rebase the PR

divyenpatel commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,438 @@
+	informer.Listen()
+	log.Infof("informer on config-map and namespaces started")
+
+	// Start routine to process pending feature state updates
```
kept it in the end, so if we encounter any error while starting config-map and namespace listeners we know, we had an issue starting the service.

divyenpatel commented:
@chethanv28 I have rebased the PR to resolve recent YAML file conflicts and also addressed all your comments.

chethanv28 commented:
Overall code changes and testing looks good to me
/approve

SandeepPissay commented on pkg/syncer/metadatasyncer.go:
```diff
@@ -168,6 +170,21 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
+			}
+			go func() {
+				if err := featurestates.StartSvFSSReplicationService(ctx, svParams.SupervisorFeatureStatesConfigInfo.Name, svParams.SupervisorFeatureStatesConfigInfo.Namespace); err != nil {
+					log.Errorf("Error starting sv FSS ReplicationService. Error: %+v", err)
```
sv -> supervisor

divyenpatel replied:
fixed.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,438 @@
+	// lock to acquire before updating namespaceUpdateMap and latestFeatureStates
+	lock *sync.RWMutex
+	// holds namespace name and status to update CR in the namespace
+	// map value true, means update is pending, false means no update is required.
```
Why do we need to interpret true or false? If namespace update is not needed why not just remove it from the map?

divyenpatel replied:
When an update or create is required for workload namespace, I am updating `namespaceUpdateMap map[string]bool` and setting true for that namespace in the map. After update/create is done, I am setting false for that namespace in the map, if failure is observed, the value remains true, so `processPendingCRUpdates`, can process the update in the next retry.

When service is starting for the first time, I am populating `namespaceUpdateMap map[string]bool` with all workload namespaces, with value true, so `processPendingCRUpdates`, can create/update CR in those namespaces.

When a new workload namespace is observed by an informer, I am adding a new namespace in this map with the value set to true, so `processPendingCRUpdates` can help create CR in the new namespace.

when a workload namespace is deleted, I am deleting an entry from `namespaceUpdateMap map[string]bool` , so processPendingCRUpdates does not need to deal pending failed updates for deleted namespace.

maintaining all current workload namespaces in the in-memory map `namespaceUpdateMap`, does not require us to query kubernetes for all workload namespaces when FSS update is observed, we can rely on namespaceUpdateMap, since it is maintained by informer events.

SandeepPissay replied:
Basically, we are not treating the problem as a producer-consumer problem. We are keeping an in-memory cache of all the current namespaces and a state whether it requires FSS replication or not. The downside of approach is more memory usage and processing all the entries in the map every time to see if any namespace needs replication.

The alternate solution could be to treat this as a producer-consume problem where the map has entries only when the FSS is updated or namespace is added. The former is very very rare and only in that case you need to query all SV namespaces.

I would prefer to optimize the solution by using less memory and less processing. But I will leave it to you as I dont want this to be a blocker.

divyenpatel replied:
> The downside of approach is more memory usage and processing all the entries in the map every time to see if any namespace needs replication.

But this comes with the advantage of reducing additional calls to the API server when config-map changes or workload namespace is added. The memory footprint is very less, as we are just storing fss config-map and workload namespaces name, and their status.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,438 @@
+	// map value true, means update is pending, false means no update is required.
+	namespaceUpdateMap map[string]bool
+	// latest featureStates
+	latestFeatureStates []featurestatesv1alpha1.FeatureState
```
Why is this required? Why cant processPendingCRUpdates query the latest FSS when it runs?

divyenpatel replied:
latestFeatureStates is loaded with the latest feature state when service starts, and we keep this up to date with config-map informer by watching on FSS config-map using informers.  This helps avoid unnecessary calls to k8s to query the latest FSS.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,438 @@
+
+	pendingCRUpdatesObj.lock.Lock()
+	defer pendingCRUpdatesObj.lock.Unlock()
+	delete(pendingCRUpdatesObj.namespaceUpdateMap, namespace.Name)
```
Looks like we are deleting the namespace from the map only on deletion of the SV namespace. Does it mean we will keep it in memory for as long as the SV namespace is present? If yes, why?

divyenpatel replied:
Yes we delete the namespace key from namespaceUpdateMap only when we observe namespace is deleted. The reason is when FSS update is required, we do not require to query, kuberetes to get list of all workload namespaces to push updates.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,438 @@
+		return err
+	}
+	for _, namespace := range workloadNamespaces.Items {
+		pendingCRUpdatesObj.namespaceUpdateMap[namespace.Name] = true
```
Now that you are listening to namespace add event, is this required?

divyenpatel replied:
We can remove this, as `func namespaceAdded(obj interface{}) {` is taking care of adding workload namespaces to `pendingCRUpdatesObj.namespaceUpdateMap` map.

divyenpatel commented:
@SandeepPissay I have addressed your review comments and validated changes with latest commit - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/764/commits/2b6e92c03718e6485d3f74b05b60d69b7439c62c

shalini-b commented:
Changes look good to me
/approve

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,418 @@
+		return err
+	}
+
+	pendingCRUpdatesObj.latestFeatureStates, err = getFeatureStates(ctx)
```
Now that you are listening to config map events, do you need this?

divyenpatel replied:
Kept it here, to fail the `StartSvFSSReplicationService` if fss config-map is not present.

SandeepPissay replied:
Why do you want to fail if config map is not present? Since you anyways have a listener and you are handling create event, it should be fine, right? I'm wondering if we are adding redundant code and logic here which is not necessary.

SandeepPissay replied:
Can you respond to this comment?

divyenpatel replied:
we are starting processPendingCRUpdates() as go routine, so there is a chance that `processPendingCRUpdates()` may get started before config-map listener done executing `configMapAdded`.

I am removing it from here and adding in inside processPendingCRUpdates() to handle that case.

```golang
// processPendingCRUpdates helps process pending CR updates at regular interval
func (pendingCRUpdatesObj *pendingCRUpdates) processPendingCRUpdates() {
	ticker := time.NewTicker(crUpdateRetryInterval)
	for range ticker.C {
		func() {
			pendingCRUpdatesObj.lock.Lock()
			defer pendingCRUpdatesObj.lock.Unlock()

			ctx, log := logger.GetNewContextWithLogger()
			var err error
			if len(pendingCRUpdatesObj.latestFeatureStates) == 0 {
				log.Warn("empty feature states observed. feature state config-map might be deleted. Trying to obtain latest feature states.")
				pendingCRUpdatesObj.latestFeatureStates, err = getFeatureStates(ctx)
				if err != nil {
					log.Errorf("Failed to get feature states. error: %v", err)
					os.Exit(1)
				}
			}
.
.
.
```

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,418 @@
+	pendingCRUpdatesObj.latestFeatureStates = featurestates
+	for namespace := range pendingCRUpdatesObj.namespaceUpdateMap {
+		pendingCRUpdatesObj.namespaceUpdateMap[namespace] = true
+		log.Infof("enqueued CR updates for workload namespace: %q", namespace)
```
Wouldn't this be too much of logging when the supervisor cluster has hundreds of namespaces? Can we move it to debug and maybe have an info log before the for-loop?

divyenpatel replied:
adding an info log after the for loop

```golang
for namespace := range pendingCRUpdatesObj.namespaceUpdateMap {
    pendingCRUpdatesObj.namespaceUpdateMap[namespace] = true
}
log.Infof("enqueued CR updates for all workload namespaces")
```

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,418 @@
+			featureState.Name = feature
+			featureState.Enabled, err = strconv.ParseBool(state)
+			if err != nil {
+				log.Warnf("failed to parse feature state value: %v for feature: %q", state, feature)
```
Why not fail?

divyenpatel replied:
kept the same logic present in the current fss check implementation based on @shalini-b 's review comments - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go#L471-L475

if we fail, then we will need to miss the entire config-map change, even when the single fss value is not correct. 
When parsing fails, setting the feature to `false` would be safe.

SandeepPissay replied:
I think we should fail so that we can catch bugs during dev.

divyenpatel replied:
ok, in that case until we fix the value of config-map parameter, we can not push the latest fss to namespaces.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,418 @@
+	}
+	if fssConfigMap.Name == supervisorFeatureStatConfigMapName &&
+		fssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		log.Warnf("supervisor feature switch state configmap %q from the namespace: %q is deleted", supervisorFeatureStatConfigMapName, supervisorFeatureStateConfigMapNamespace)
```
Dont you have to clear the map since FSS itself is not there?

divyenpatel replied:
If we clear the map, then after fss config-map is deleted accidentally, we will end up setting `false` in all existing workload namespaces and all new workload namespaces added after deleting fss config-map will have all features set as false. So just kept this as a warning message. 

Ideally, SV admin should never delete this internal config-map.

SandeepPissay replied:
If FSS config map is deleted, there is no need to perform any replication. We can keep the logic simple by just clearing the map.

divyenpatel replied:
updating configMapDeleted as below to clear in-memory cached feature switch states.

```
// configMapDeleted is called when config-map is deleted
func configMapDeleted(obj interface{}) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	ctx = logger.NewContextWithLogger(ctx)
	log := logger.GetLogger(ctx)

	fssConfigMap, ok := obj.(*v1.ConfigMap)
	if fssConfigMap == nil || !ok {
		log.Warnf("configMapDeleted: unrecognized object %+v", obj)
		return
	}
	if fssConfigMap.Name == supervisorFeatureStatConfigMapName &&
		fssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
		log.Warnf("supervisor feature switch state configmap %q from the namespace: %q is deleted", supervisorFeatureStatConfigMapName, supervisorFeatureStateConfigMapNamespace)
		pendingCRUpdatesObj.lock.Lock()
		defer pendingCRUpdatesObj.lock.Unlock()
		pendingCRUpdatesObj.latestFeatureStates = make([]featurestatesv1alpha1.FeatureState, 0)
		for namespace := range pendingCRUpdatesObj.namespaceUpdateMap {
			pendingCRUpdatesObj.namespaceUpdateMap[namespace] = false
		}
	}
}
```

divyenpatel replied:
Also updating processPendingCRUpdates() to handle case when empty fss is observed.

```golang
// processPendingCRUpdates helps process pending CR updates at regular interval
func (pendingCRUpdatesObj *pendingCRUpdates) processPendingCRUpdates() {
	ticker := time.NewTicker(crUpdateRetryInterval)
	for range ticker.C {
		func() {
			pendingCRUpdatesObj.lock.Lock()
			defer pendingCRUpdatesObj.lock.Unlock()

			ctx, log := logger.GetNewContextWithLogger()
			if len(pendingCRUpdatesObj.latestFeatureStates) == 0 {
				log.Warnf("empty feature states observed. feature state config-map might be deleted.")
				return
			}
.
.

```

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,418 @@
+	_, labelPresentInOldNamespace := oldNamespace.Labels[WorkLoadNamespaceLabelKey]
+	_, labelPresentInNewNamespace := newNamespace.Labels[WorkLoadNamespaceLabelKey]
+
+	if !labelPresentInOldNamespace && labelPresentInNewNamespace {
```
All we care about is whether the label is present in the latest namespace object right?

divyenpatel replied:
yes so when namespace is updated, if old namespace object does not have workload namespace label, but new namespace object has workload namespace label, means namespace is now being used as workload namespace, so adding it to the map.

for all other updates, `namespaceUpdated` will have no-op.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -263,7 +267,8 @@ func configMapAdded(obj interface{}) {
 			if err != nil {
-				log.Warnf("failed to parse feature state value: %v for feature: %q", state, feature)
+				log.Errorf("failed to parse feature state value: %v for feature: %q", state, feature)
+				return
```
Since this is a issue that can occur at dev time, can we invoke os.Exit to crash the container?

SandeepPissay replied:
On more thinking, I think we should crash CSI once the FSS has invalid entries or the FSS itself is deleted. Silently ignoring the problem will be problematic.

divyenpatel replied:
crashing the container is a good idea. so during dev and fvt testing, we can identify the issue earlier.
making the following change.

```
                       featureState.Enabled, err = strconv.ParseBool(state)
			if err != nil {
				log.Errorf("failed to parse feature state value: %v for feature: %q", state, feature)
				os.Exit(1)
			}
```

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,430 @@
+			defer pendingCRUpdatesObj.lock.Unlock()
+
+			ctx, log := logger.GetNewContextWithLogger()
+			if len(pendingCRUpdatesObj.latestFeatureStates) == 0 {
```
Does this mean, we may have stale replicated FSS config maps in all supervisor namespaces? If yes, how can we solve this? Since we do not expect anyone to delete the supervisor FSS config map, we should be fine. I'm worried about a code change that deletes the supervisor config map in the future. The current code change is silently ignoring the problem. I would suggest to crash the container so that we catch such problems early during dev. Basically if there is no supervisor FSS or if it is deleted during runtime, we crash the CSI container and CSI will never come up. This is a nice way to get informed about such issues early. Thoughts?

divyenpatel replied:
Addressed this with https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/764#discussion_r606515866

divyenpatel commented:
squashed all commits to prepare PR for merge

divyenpatel commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,427 @@
+			namespaceDeleted(obj)
+		})
+	informer.Listen()
+	log.Infof("informer on config-map and namespaces started")
```
fixed at all places in `pkg/internalapis/featurestates/featurestates.go`

divyenpatel commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,427 @@
+
+	// Start routine to process pending feature state updates
+	go pendingCRUpdatesObj.processPendingCRUpdates()
+	log.Infof("started background routine to process pending feature state updates at regular interval")
```
fixed.

divyenpatel commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,427 @@
+				pendingCRUpdatesObj.latestFeatureStates, err = getFeatureStates(ctx)
+				if err != nil {
+					log.Errorf("Failed to get feature states. error: %v", err)
+					os.Exit(1)
```
done updated code as below in `processPendingCRUpdates`.

```golang
if len(pendingCRUpdatesObj.latestFeatureStates) == 0 {
				log.Warn("empty feature states observed. feature state config-map might be deleted. Trying to obtain latest feature states.")
				pendingCRUpdatesObj.latestFeatureStates, err = getFeatureStates(ctx)
				if err != nil {
					log.Errorf("failed to get feature states. error: %v", err)
					return
				}
			}
```

updated `getFeatureStates` to exit container when 1) not found error observed, and  2) parsing fails.

```golang
func getFeatureStates(ctx context.Context) ([]featurestatesv1alpha1.FeatureState, error) {
	log := logger.GetLogger(ctx)
	//  Retrieve SV FeatureStates configmap
	fssConfigMap, err := k8sClient.CoreV1().ConfigMaps(supervisorFeatureStateConfigMapNamespace).Get(ctx, supervisorFeatureStatConfigMapName, metav1.GetOptions{})
	if err != nil {
		if apierrors.IsNotFound(err) {
			log.Errorf("supervisor feature switch state with name: %q not found in the namespace: %q", supervisorFeatureStatConfigMapName, supervisorFeatureStateConfigMapNamespace)
			os.Exit(1)
		}
		log.Errorf("failed to retrieve SV feature switch state from namespace:%q with name: %q", supervisorFeatureStateConfigMapNamespace, supervisorFeatureStatConfigMapName)
		return nil, err
	}

	log.Infof("Successfully retrieved SV feature switch state from namespace:%q with name: %q", supervisorFeatureStateConfigMapNamespace, supervisorFeatureStatConfigMapName)
	var featureStates []featurestatesv1alpha1.FeatureState
	for feature, state := range fssConfigMap.Data {
		var featureState featurestatesv1alpha1.FeatureState
		featureState.Name = feature
		featureState.Enabled, err = strconv.ParseBool(state)
		if err != nil {
			log.Errorf("failed to parse feature state value: %v for feature: %q. Defaulting to false", state, feature)
			os.Exit(1)
		}
		featureStates = append(featureStates, featureState)
	}
	return featureStates, nil
}

```

divyenpatel commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,427 @@
+					log.Debugf("CR update is not required for the namespace: %q", namespace)
+					continue
+				}
+				log.Infof("feature state update is required for namespace: %q", namespace)
```
fixed.

divyenpatel commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,427 @@
+						err = controllerRuntimeClient.Create(ctx, cnsCsiSvFeatureStates)
+						if err != nil {
+							log.Errorf("failed to create cnsCsiSvFeatureStates CR instance in the "+
+								"namespace: %q, Err: %v", namespace, err)
```
done.

divyenpatel commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -0,0 +1,427 @@
+							continue
+						}
+						pendingCRUpdatesObj.namespaceUpdateMap[namespace] = false
+						log.Infof("created cnsCsiSvFeatureStates CR instance in the namespace: %q", namespace)
```
fixed.

SandeepPissay commented on pkg/internalapis/featurestates/featurestates.go:
```diff
@@ -407,19 +407,23 @@ func getFeatureStates(ctx context.Context) ([]featurestatesv1alpha1.FeatureState
 		featureState.Name = feature
 		featureState.Enabled, err = strconv.ParseBool(state)
 		if err != nil {
 			log.Errorf("failed to parse feature state value: %v for feature: %q. Defaulting to false", state, feature)
```
Can you remove "Defaulting to false" in the error log?

divyenpatel replied:
fixed.

SandeepPissay commented:
Looks good to me. Can you restart the E2E pipelines to make sure the latest code is not breaking existing tests?

/approve

divyenpatel commented:
jtest wcp
jtest gc

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
Started GC block pipeline...

svcbot-qecnsdp commented:
``` 
GC build status: FAILURE 
Stage before exit: deploy-gc-testbed 

```

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 31 of 166 Specs in 12112.232 seconds
FAIL! -- 30 Passed | 1 Failed | 0 Pending | 135 Skipped
--- FAIL: TestE2E (12112.34s)
FAIL

Ginkgo ran 1 suite in 3h23m45.465122682s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver`

```

chethanv28 commented:
jtest gc

chethanv28 commented:
> ```
> WCP build status: FAILURE 
> Stage before exit: null 
> Jenkins E2E Test Results: 
> Ran 31 of 166 Specs in 12112.232 seconds
> FAIL! -- 30 Passed | 1 Failed | 0 Pending | 135 Skipped
> --- FAIL: TestE2E (12112.34s)
> FAIL
> 
> Ginkgo ran 1 suite in 3h23m45.465122682s
> Test Suite Failed
> make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver`
> ```

@divyenpatel Can you confirm what is the 1 e2e test failure is about

divyenpatel commented:
> > ```
> > WCP build status: FAILURE 
> > Stage before exit: null 
> > Jenkins E2E Test Results: 
> > Ran 31 of 166 Specs in 12112.232 seconds
> > FAIL! -- 30 Passed | 1 Failed | 0 Pending | 135 Skipped
> > --- FAIL: TestE2E (12112.34s)
> > FAIL
> > 
> > Ginkgo ran 1 suite in 3h23m45.465122682s
> > Test Suite Failed
> > make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver`
> > ```
> 
> @divyenpatel Can you confirm what is the 1 e2e test failure ?

The issue is not related to this change. Filing internal issue for further investigation.

```
00:23:17  Summarizing 1 Failure:
00:23:17  
00:23:17  [Fail] [csi-block-vanilla] full-sync-test [It] [csi-supervisor] Verify labels are created in CNS after updating pvc and/or pv with new labels 
00:23:17  /home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver/tests/e2e/util.go:315
00:23:17  
00:23:17  Ran 31 of 166 Specs in 12112.232 seconds
00:23:17  FAIL! -- 30 Passed | 1 Failed | 0 Pending | 135 Skipped
00:23:17  --- FAIL: TestE2E (12112.34s)
00:23:17  FAIL
```

```
21:15:14  • Failure [112.898 seconds]
21:15:14  [csi-block-vanilla] full-sync-test
21:15:14  /home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver/tests/e2e/fullsync_test_for_block_volume.go:55
21:15:14    [csi-supervisor] Verify labels are created in CNS after updating pvc and/or pv with new labels [It]
21:15:14    /home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver/tests/e2e/fullsync_test_for_block_volume.go:198
21:15:14  
21:15:14    Failed to create pvc with err: PVC Create API error: Post https://10.186.41.111:6443/api/v1/namespaces/e2e-test-namespace/persistentvolumeclaims: stream error: stream ID 377; INTERNAL_ERROR
21:15:14    Unexpected error:
21:15:14        <*errors.errorString | 0xc0007d9730>: {
21:15:14            s: "PVC Create API error: Post https://10.186.41.111:6443/api/v1/namespaces/e2e-test-namespace/persistentvolumeclaims: stream error: stream ID 377; INTERNAL_ERROR",
21:15:14        }
21:15:14        PVC Create API error: Post https://10.186.41.111:6443/api/v1/namespaces/e2e-test-namespace/persistentvolumeclaims: stream error: stream ID 377; INTERNAL_ERROR
21:15:14    occurred
21:15:14  
21:15:14    /home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver/tests/e2e/util.go:315
21:15:14  
21:15:14    Full Stack Trace
21:15:14    sigs.k8s.io/vsphere-csi-driver/tests/e2e.createPVC(0x2b3dc60, 0xc00088e420, 0xc00005428e, 0x12, 0x0, 0x0, 0x0, 0xc0007a3680, 0x0, 0x0, ...)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver/tests/e2e/util.go:315 +0x402
21:15:14    sigs.k8s.io/vsphere-csi-driver/tests/e2e.createPVCAndStorageClass(0x2b3dc60, 0xc00088e420, 0xc00005428e, 0x12, 0x0, 0xc0007f29f0, 0x0, 0x0, 0x0, 0x0, ...)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver/tests/e2e/util.go:293 +0x24b
21:15:14    sigs.k8s.io/vsphere-csi-driver/tests/e2e.glob..func6.4()
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver/tests/e2e/fullsync_test_for_block_volume.go:216 +0x1d79
21:15:14    github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc0003eefc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/leafnodes/runner.go:113 +0xb8
21:15:14    github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc0003eefc0, 0xc000056538, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/leafnodes/runner.go:64 +0xcf
21:15:14    github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run(0xc0003da980, 0x2a8e160, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/leafnodes/it_node.go:26 +0x64
21:15:14    github.com/onsi/ginkgo/internal/spec.(*Spec).runSample(0xc0002061e0, 0x0, 0x2a8e160, 0xc0000a6a00)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/spec/spec.go:215 +0x5b5
21:15:14    github.com/onsi/ginkgo/internal/spec.(*Spec).Run(0xc0002061e0, 0x2a8e160, 0xc0000a6a00)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/spec/spec.go:138 +0x101
21:15:14    github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec(0xc0000e54a0, 0xc0002061e0, 0x0)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/specrunner/spec_runner.go:200 +0x10f
21:15:14    github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs(0xc0000e54a0, 0x1)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/specrunner/spec_runner.go:170 +0x120
21:15:14    github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run(0xc0000e54a0, 0xc0004807d0)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/specrunner/spec_runner.go:66 +0x117
21:15:14    github.com/onsi/ginkgo/internal/suite.(*Suite).Run(0xc000102af0, 0x7f4e549df948, 0xc000100100, 0x26c7bee, 0x1f, 0xc00052bb70, 0x1, 0x1, 0x2aef2c0, 0xc0000a6a00, ...)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/internal/suite/suite.go:79 +0x586
21:15:14    github.com/onsi/ginkgo.runSpecsWithCustomReporters(0x2a92440, 0xc000100100, 0x26c7bee, 0x1f, 0xc000097f18, 0x1, 0x1, 0xc000097f28)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/ginkgo_dsl.go:238 +0x217
21:15:14    github.com/onsi/ginkgo.RunSpecs(0x2a92440, 0xc000100100, 0x26c7bee, 0x1f, 0xc0000758b0)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/go/pkg/mod/github.com/onsi/ginkgo@v1.16.0/ginkgo_dsl.go:213 +0xa7
21:15:14    sigs.k8s.io/vsphere-csi-driver/tests/e2e.TestE2E(0xc000100100)
21:15:14    	/home/worker/workspace/github-csi-wcp-CICD/Results/568/vsphere-csi-driver/tests/e2e/suite_test.go:50 +0x9c
21:15:14    testing.tRunner(0xc000100100, 0x27992e8)
21:15:14    	/usr/local/go/src/testing/testing.go:909 +0xc9
21:15:14    created by testing.(*T).Run
21:15:14    	/usr/local/go/src/testing/testing.go:960 +0x350
```

chethanv28 commented:
Code changes and testing looks good to me. The new code added by the PR is within clusterFlavor check i.e., Supervisor Cluster.
Thanks for addressing all the comments @divyenpatel 
/lgtm

svcbot-qecnsdp commented:
``` 
GC build status: FAILURE 
Stage before exit: replace-image-in-vc 

```
